server/src/main/java/org/elasticsearch/cluster/project/ProjectStateRegistry.java

```diff
@@ -95,12 +118,44 @@ public TransportVersion getMinimalSupportedVersion() {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeMap(projectsSettings);
+        if (out.getTransportVersion().onOrAfter(TransportVersions.PROJECT_STATE_REGISTRY_RECORDS_DELETIONS)) {
+            out.writeCollection(projectsMarkedForDeletion);
+            out.writeZLong(projectsMarkedForDeletionGeneration);
+        } else {
+            // There should be no deletion unless all MP nodes are at or after PROJECT_STATE_REGISTRY_RECORDS_DELETIONS
+            assert projectsMarkedForDeletion.isEmpty();
+            assert projectsMarkedForDeletionGeneration == 0;
+        }
     }
 
     public int size() {
         return projectsSettings.size();
     }
 
+    public long getProjectsMarkedForDeletionGeneration() {
+        return projectsMarkedForDeletionGeneration;
+    }
+
+    // visible for testing
+    Map<ProjectId, Settings> getProjectsSettings() {
+        return Collections.unmodifiableMap(projectsSettings);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o instanceof ProjectStateRegistry == false) return false;
+        ProjectStateRegistry that = (ProjectStateRegistry) o;
+        return projectsMarkedForDeletionGeneration == that.projectsMarkedForDeletionGeneration
+            && Objects.equals(projectsSettings, that.projectsSettings)
+            && Objects.equals(projectsMarkedForDeletion, that.projectsMarkedForDeletion);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(projectsSettings, projectsMarkedForDeletion, projectsMarkedForDeletionGeneration);
+    }
```

Comment:
Many of the objects in cluster state compares equality by identity, e.g. `DiscoveryNodes`, so that it is a bit more performant when [there is no change](https://github.com/elastic/elasticsearch/blob/cacce50192ac29e1aaca2e75c170a90d15af292c/server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java#L535-L537) which is often the case. So I'd prefer we use identity equality here if these methods are only used for tests and use a [wrapper class](https://github.com/elastic/elasticsearch/blob/cacce50192ac29e1aaca2e75c170a90d15af292c/server/src/test/java/org/elasticsearch/cluster/block/ClusterBlocksSerializationTests.java#L281-L287) in tests.

Reply:
I wasted a bit of time trying to make it happen before realizing how ugly it can get since this is a bit more than ClusterBlocks and would require more hacking. In fact, `SnapshotsInProgress` which is a more similar case to this (also a custom) and does have serialization tests, also does what I've done here.

Reply:
(same with HealthMetadata, RestoreInProgress, to name a few other customs)

Reply:
OK. There is indeed inconsistency in how we handle equals for cluster state objects. Since this object is linear to number of projects, the upper-bound should not be too high for performing a non-identity equals. We can also revisit it if it proves to be a problem.

Reply:
If we make this class to support more granular diff than the generic [CompleteNameDiff](https://github.com/elastic/elasticsearch/blob/a69c48477f2fbca93115613c2b512e8b469142e3/server/src/main/java/org/elasticsearch/cluster/AbstractNamedDiffable.java#L28-L31). The identity equality will be much less interesting since I think creating diff is one place `equals` is called the most often. That is the case for `SnapshotsInProgress`, but not `RestoreInProgress`, again inconsistency. In any case, it's not for this PR.